pkg/cri/resource-manager/policy/builtin/podpools/podpools-policy.go

```diff
@@ -0,0 +1,727 @@
+	}
+	return nil
+}
+
+// ReleaseResources is a resource release request for this policy.
+func (p *podpools) ReleaseResources(c cache.Container) error {
+	log.Debug("releasing container %s...", c.PrettyName())
+	pod, ok := c.GetPod()
+	if !ok {
+		return podpoolsError("cannot find pod of container %s from the cache", c.PrettyName())
+	}
+	if pool := p.allocatedPool(pod); pool != nil {
+		p.dismissContainer(c, pool)
+		if log.DebugEnabled() {
+			log.Debug(p.dumpPool(pool))
+		}
+		if p.containersInPool(pod, pool) == 0 {
+			log.Debug("all containers removed, free pool allocation %s for pod %q", pool.PrettyName(), pod.GetName())
+			p.validatePodCPU(pod, pool)
+			p.freePool(pod, pool)
+		}
+	} else {
+		log.Debug("ReleaseResources: pool-less container %s, nothing to release", c.PrettyName())
+	}
+	return nil
+}
+
+// UpdateResources is a resource allocation update request for this policy.
+func (p *podpools) UpdateResources(c cache.Container) error {
+	log.Debug("(not) updating container %s...", c.PrettyName())
+	return nil
+}
+
+// Rebalance tries to find an optimal allocation of resources for the current containers.
+func (p *podpools) Rebalance() (bool, error) {
+	log.Debug("(not) rebalancing containers...")
+	return false, nil
+}
+
+// HandleEvent handles policy-specific events.
+func (p *podpools) HandleEvent(*events.Policy) (bool, error) {
+	log.Debug("(not) handling event...")
+	return false, nil
+}
+
+// ExportResourceData provides resource data to export for the container.
+func (p *podpools) ExportResourceData(c cache.Container) map[string]string {
+	return nil
+}
+
+// Introspect provides data for external introspection.
+func (p *podpools) Introspect(*introspect.State) {
+	return
+}
+
+// allocatedPool returns a pool already allocated for a pod.
+func (p *podpools) allocatedPool(pod cache.Pod) *Pool {
+	podID := pod.GetID()
+	pools := filterPools(p.pools,
+		func(pl *Pool) bool { _, ok := pl.PodIDs[podID]; return ok })
+	if len(pools) == 0 {
+		return nil
+	}
+	return pools[0]
+}
+
+// allocatePool returns a pool allocated for a pod.
+func (p *podpools) allocatePool(pod cache.Pod) *Pool {
+	if pool := p.allocatedPool(pod); pool != nil {
+		return pool
+	}
+	poolDef := p.getPoolDef(pod)
+	if poolDef == nil {
+		return nil
+	}
+	// Try to find a suitable pool and allocate it for the pod.
+	pools := filterPools(p.pools,
+		func(pl *Pool) bool {
+			return poolDef.Name == pl.Def.Name && (pl.Def.MaxPods > len(pl.PodIDs) || pl.Def.MaxPods == 0)
+		})
+	// Sort pools according to pool type fill order so that the
+	// first pool in the list is the preferred one.
+	switch poolDef.FillOrder {
+	case FillBalanced:
+		sort.Slice(pools, func(i, j int) bool {
+			return len(pools[i].PodIDs) < len(pools[j].PodIDs)
+		})
+	case FillPacked:
+		sort.Slice(pools, func(i, j int) bool {
+			return len(pools[i].PodIDs) > len(pools[j].PodIDs)
+		})
```

Comment:
That's very correct. I considered logging a warning or an error message, but finally ended up rejecting the pool configuration with "_n_ pool(s) unreachable..." error message. Still feeling a bit uncertain as I tend to avoid writing `if`s that only prevent behavior, even shooting oneself in the foot. Maybe this time it's not just my lack of imagination that I can't come up with a use case for unreachable pools... :)

Reply:
> Still feeling a bit uncertain as I tend to avoid writing `if`s that only prevent behavior, even shooting oneself in the foot

Yeah, that is a good point and a good principle – too "smart" software can be a real pita. I'm counting on your judgement here 😉